File: src/engine/adapters/kafka/definition.ts
import type { KafkaJS } from '@confluentinc/kafka-javascript';
import { ConsumerGroupStates, ConsumerGroupTypes } from '@confluentinc/kafka-javascript';
import type { DefinitionSection, ObjectDefinition } from '../../../shared/domain/definition';
import { encodePath } from '../../../shared/domain/tree';
import { AdapterError } from '../errors';
import { mapKafkaError } from './errors';

// P23 D5: where a topic's/group's partition and member data went once the tree stopped showing it
// (P19's "relocation, not deletion" ground rule, F4). Every value here is already typed in the
// installed client — no new dependency, no new round trip beyond what the user's own open costs.

// P32 D15: this client's GroupDescription types state/type as numeric enums
// (ConsumerGroupStates/ConsumerGroupTypes), not kafkajs's plain strings — rendering the number
// would be a regression, so both get a name lookup rather than `String(state)`. The runtime
// objects behind these `.d.ts` enum declarations are plain one-way name->number maps (no reverse
// mapping the way a real TS-compiled enum would have — verified against lib/admin.js), so the
// reverse lookup is built here rather than indexed directly.
function reverseLookup(source: Record<string, unknown>): Map<number, string> {
  const map = new Map<number, string>();
  for (const [name, value] of Object.entries(source)) {
    if (typeof value === 'number') map.set(value, name);
  }
  return map;
}
const STATE_NAMES = reverseLookup(ConsumerGroupStates);
const TYPE_NAMES = reverseLookup(ConsumerGroupTypes);

function groupStateName(state: ConsumerGroupStates): string {
  return STATE_NAMES.get(state) ?? `unknown (${state})`;
}
function groupTypeName(type: ConsumerGroupTypes): string {
  return TYPE_NAMES.get(type) ?? `unknown (${type})`;
}

export async function buildTopicDefinition(
  admin: KafkaJS.Admin,
  topic: string,
): Promise<ObjectDefinition> {
  let topics: KafkaJS.ITopicMetadata[];
  try {
    topics = await admin.fetchTopicMetadata({ topics: [topic] });
  } catch (err) {
    throw mapKafkaError(err);
  }
  const partitions = (topics.find((t) => t.name === topic)?.partitions ?? [])
    .slice()
    .sort((a, b) => a.partitionId - b.partitionId);

  const partitionsSection: DefinitionSection = {
    title: 'Partitions',
    rows: partitions.map((p) => ({
      name: String(p.partitionId),
      value: `leader ${p.leader}`,
      detail: `replicas ${p.replicas.join(',')} · isr ${p.isr.join(',')}`,
    })),
  };

  // P32 D14/F11: this client has no describeConfigs and no describeCluster — not on the compat
  // Admin, not on the native AdminClient either (grep describeconfig/clusterid over the package
  // finds nothing but two prose mentions of the broker-side API). The section stays, rendered
  // empty, with a note saying why — the same degradation shape the old DESCRIBE_CONFIGS-denied
  // path already used (a missing section must not fail the whole tab), just permanent now rather
  // than ACL-dependent. librdkafka's C API has rd_kafka_DescribeConfigs; this binding simply does
  // not wrap it, so this can come back from upstream without any change here.
  const configSection: DefinitionSection = { title: 'Configuration', rows: [] };
  const notes: string[] = [
    'Topic configuration is not available: this Kafka client has no DescribeConfigs call.',
  ];

  const doc = {
    partitions: partitions.map((p) => ({
      id: p.partitionId,
      leader: p.leader,
      replicas: p.replicas,
      isr: p.isr,
    })),
    config: configSection.rows.map((r) => ({
      name: r.name,
      value: r.value,
      default: r.detail === 'default',
    })),
  };

  return {
    path: encodePath([{ kind: 'topic', name: topic }]),
    kind: 'topic',
    qualifiedName: topic,
    language: 'json',
    statements: [JSON.stringify(doc, null, 2)],
    origin: 'server',
    notes,
    constraints: [],
    documentSchema: null,
    sections: [partitionsSection, configSection],
    generatedAt: new Date().toISOString(),
  };
}

export async function buildGroupDefinition(
  admin: KafkaJS.Admin,
  groupId: string,
): Promise<ObjectDefinition> {
  // GroupDescription isn't itself re-exported from the kafkajs compat surface (only the plural
  // GroupDescriptions is) — pull the element type out rather than reaching into ./rdkafka directly.
  let group: KafkaJS.GroupDescriptions['groups'][number] | undefined;
  try {
    ({
      groups: [group],
    } = await admin.describeGroups([groupId]));
  } catch (err) {
    throw mapKafkaError(err);
  }
  if (!group) throw new AdapterError('E_NOT_FOUND', `consumer group not found: ${groupId}`);

  // P32 D15: partitionAssignor and coordinator are new in this client's GroupDescription — worth
  // a row each, since a Kafka 4 client's group view is exactly where "classic vs KIP-848 consumer
  // protocol" and "which broker owns this group" belong.
  const groupSection: DefinitionSection = {
    title: 'Group',
    rows: [
      { name: 'state', value: groupStateName(group.state), detail: null },
      { name: 'type', value: groupTypeName(group.type), detail: null },
      { name: 'protocolType', value: group.protocolType || '—', detail: null },
      { name: 'protocol', value: group.protocol || '—', detail: null },
      { name: 'partitionAssignor', value: group.partitionAssignor || '—', detail: null },
      {
        name: 'coordinator',
        value: `${group.coordinator.host}:${group.coordinator.port}`,
        detail: null,
      },
      { name: 'members', value: String(group.members.length), detail: null },
    ],
  };

  const membersSection: DefinitionSection = {
    title: 'Members',
    rows: group.members.map((m) => ({
      name: m.clientId,
      value: m.clientHost,
      detail: m.memberId,
    })),
  };

  // fetchOffsets is a second, independent call — a group with read access but no offset-fetch
  // permission still shows its Group/Members sections rather than failing the whole load.
  let offsetsSection: DefinitionSection = { title: 'Committed offsets', rows: [] };
  const notes: string[] = [];
  try {
    const offsets = await admin.fetchOffsets({ groupId });
    offsetsSection = {
      title: 'Committed offsets',
      rows: offsets.flatMap((t) =>
        t.partitions
          .slice()
          .sort((a, b) => a.partition - b.partition)
          .map((p) => ({ name: `${t.topic}[${p.partition}]`, value: p.offset, detail: null })),
      ),
    };
  } catch {
    notes.push('Committed offsets could not be read.');
  }

  // P32 D15: doc mirrors groupSection's human-readable values for state/type, not the raw numeric
  // enum — unlike a topic's partition leader/replica ids (already meaningful integers), a bare
  // state/type number in the JSON statements view would be opaque without cross-referencing
  // ConsumerGroupStates/ConsumerGroupTypes.
  const doc = {
    state: groupStateName(group.state),
    type: groupTypeName(group.type),
    protocolType: group.protocolType,
    protocol: group.protocol,
    partitionAssignor: group.partitionAssignor,
    coordinator: { host: group.coordinator.host, port: group.coordinator.port },
    members: group.members.map((m) => ({
      clientId: m.clientId,
      clientHost: m.clientHost,
      memberId: m.memberId,
    })),
    offsets: offsetsSection.rows,
  };

  return {
    path: encodePath([{ kind: 'consumerGroup', name: groupId }]),
    kind: 'consumerGroup',
    qualifiedName: groupId,
    language: 'json',
    statements: [JSON.stringify(doc, null, 2)],
    origin: 'server',
    notes,
    constraints: [],
    documentSchema: null,
    sections: [groupSection, membersSection, offsetsSection],
    generatedAt: new Date().toISOString(),
  };
}

File: apps/kira-studio/internal/adapters/kafka/definition.go
package kafka

import (
	"context"
	"encoding/json"
	"fmt"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/twmb/franz-go/pkg/kadm"
	"github.com/twmb/franz-go/pkg/kmsg"

	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/adapters"
	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/storage/model"
)

func joinInts(vals []int32) string {
	parts := make([]string, len(vals))
	for i, v := range vals {
		parts[i] = strconv.FormatInt(int64(v), 10)
	}
	return strings.Join(parts, ",")
}

type topicPartitionDoc struct {
	ID       int32   `json:"id"`
	Leader   int32   `json:"leader"`
	Replicas []int32 `json:"replicas"`
	ISR      []int32 `json:"isr"`
}

type topicConfigDoc struct {
	Name    string `json:"name"`
	Value   string `json:"value"`
	Default bool   `json:"default"`
}

// buildTopicDefinition is definition.ts's buildTopicDefinition (:35-97). P58e E11: the
// Configuration section is populated via kadm.DescribeTopicConfigs — the capability the
// TypeScript adapter's own comment recorded as permanently lost, because "this binding simply
// does not wrap [DescribeConfigs]" — replacing the permanent "not available" note entirely.
func buildTopicDefinition(ctx context.Context, adm *kadm.Client, topic string) (model.ObjectDefinition, error) {
	meta, err := adm.Metadata(ctx, topic)
	if err != nil {
		return model.ObjectDefinition{}, mapError(err)
	}
	var partitions []kadm.PartitionDetail
	if detail, ok := meta.Topics[topic]; ok {
		partitions = detail.Partitions.Sorted()
	}

	partitionsSection := model.DefinitionSection{Title: "Partitions", Rows: make([]model.DefinitionSectionRow, len(partitions))}
	partitionDocs := make([]topicPartitionDoc, len(partitions))
	for i, p := range partitions {
		detail := fmt.Sprintf("replicas %s · isr %s", joinInts(p.Replicas), joinInts(p.ISR))
		partitionsSection.Rows[i] = model.DefinitionSectionRow{
			Name: strconv.FormatInt(int64(p.Partition), 10), Value: fmt.Sprintf("leader %d", p.Leader), Detail: &detail,
		}
		partitionDocs[i] = topicPartitionDoc{ID: p.Partition, Leader: p.Leader, Replicas: p.Replicas, ISR: p.ISR}
	}

	configSection := model.DefinitionSection{Title: "Configuration", Rows: []model.DefinitionSectionRow{}}
	var configDocs []topicConfigDoc
	var notes []string

	resourceConfigs, cfgErr := adm.DescribeTopicConfigs(ctx, topic)
	rc, onErr := resourceConfigs.On(topic, nil)
	switch {
	case cfgErr != nil, onErr != nil, rc.Err != nil:
		// A missing section must not fail the whole tab — the same degradation shape the
		// ACL-denied case already needs, just no longer the permanent state.
		notes = append(notes, "Topic configuration could not be read.")
	default:
		configs := append([]kadm.Config(nil), rc.Configs...)
		sort.Slice(configs, func(i, j int) bool { return configs[i].Key < configs[j].Key })
		configSection.Rows = make([]model.DefinitionSectionRow, len(configs))
		configDocs = make([]topicConfigDoc, len(configs))
		for i, c := range configs {
			isDefault := c.Source == kmsg.ConfigSourceDefaultConfig
			detail := c.Source.String()
			if isDefault {
				detail = "default"
			}
			value := c.MaybeValue()
			configSection.Rows[i] = model.DefinitionSectionRow{Name: c.Key, Value: value, Detail: &detail}
			configDocs[i] = topicConfigDoc{Name: c.Key, Value: value, Default: isDefault}
		}
	}

	doc := struct {
		Partitions []topicPartitionDoc `json:"partitions"`
		Config     []topicConfigDoc    `json:"config"`
	}{Partitions: partitionDocs, Config: configDocs}
	statement, err := json.MarshalIndent(doc, "", "  ")
	if err != nil {
		return model.ObjectDefinition{}, mapError(err)
	}

	return model.ObjectDefinition{
		Path:           model.EncodePath([]model.PathSegment{{Kind: "topic", Name: topic}}),
		Kind:           "topic",
		QualifiedName:  topic,
		Language:       "json",
		Statements:     []string{string(statement)},
		Origin:         "server",
		Notes:          notes,
		Constraints:    []model.ConstraintMeta{},
		DocumentSchema: nil,
		Sections:       []model.DefinitionSection{partitionsSection, configSection},
		GeneratedAt:    time.Now().UTC().Format(time.RFC3339Nano),
	}, nil
}

type groupMemberDoc struct {
	ClientID   string `json:"clientId"`
	ClientHost string `json:"clientHost"`
	MemberID   string `json:"memberId"`
}

type groupOffsetDoc struct {
	Name  string `json:"name"`
	Value string `json:"value"`
}

// buildGroupDefinition is definition.ts's buildGroupDefinition (:99-195). P58e E13:
// kadm.DescribedGroup has no Type field and no PartitionAssignor field — the group definition
// drops its `type` row entirely (there is nothing in kadm to reverse-lookup a numeric enum from,
// unlike definition.ts:12-33's whole apparatus) and merges `partitionAssignor` into `protocol`,
// since kadm's own doc names Protocol "the partition assignor strategy this group is using" — the
// same value under a second name. State is already a string on this client, so the group section
// drops from seven rows to five: state, protocolType, protocol, coordinator, members.
func buildGroupDefinition(ctx context.Context, adm *kadm.Client, groupID string) (model.ObjectDefinition, error) {
	groups, err := adm.DescribeGroups(ctx, groupID)
	if err != nil {
		return model.ObjectDefinition{}, mapError(err)
	}
	group, ok := groups[groupID]
	if !ok || group.Err != nil {
		return model.ObjectDefinition{}, adapters.New(adapters.CodeNotFound, "consumer group not found: "+groupID, nil)
	}

	orEmDash := func(s string) string {
		if s == "" {
			return "—"
		}
		return s
	}
	coordinator := fmt.Sprintf("%s:%d", group.Coordinator.Host, group.Coordinator.Port)
	groupSection := model.DefinitionSection{
		Title: "Group",
		Rows: []model.DefinitionSectionRow{
			{Name: "state", Value: orEmDash(group.State), Detail: nil},
			{Name: "protocolType", Value: orEmDash(group.ProtocolType), Detail: nil},
			{Name: "protocol", Value: orEmDash(group.Protocol), Detail: nil},
			{Name: "coordinator", Value: coordinator, Detail: nil},
			{Name: "members", Value: strconv.Itoa(len(group.Members)), Detail: nil},
		},
	}

	membersSection := model.DefinitionSection{Title: "Members", Rows: make([]model.DefinitionSectionRow, len(group.Members))}
	memberDocs := make([]groupMemberDoc, len(group.Members))
	for i, m := range group.Members {
		memberID := m.MemberID
		membersSection.Rows[i] = model.DefinitionSectionRow{Name: m.ClientID, Value: m.ClientHost, Detail: &memberID}
		memberDocs[i] = groupMemberDoc{ClientID: m.ClientID, ClientHost: m.ClientHost, MemberID: m.MemberID}
	}

	// A second, independent call — a group with read access but no offset-fetch permission still
	// shows its Group/Members sections rather than failing the whole load.
	offsetsSection := model.DefinitionSection{Title: "Committed offsets", Rows: []model.DefinitionSectionRow{}}
	var offsetDocs []groupOffsetDoc
	var notes []string
	if offsets, err := adm.FetchOffsets(ctx, groupID); err != nil {
		notes = append(notes, "Committed offsets could not be read.")
	} else {
		sorted := offsets.Sorted()
		offsetsSection.Rows = make([]model.DefinitionSectionRow, len(sorted))
		offsetDocs = make([]groupOffsetDoc, len(sorted))
		for i, o := range sorted {
			name := fmt.Sprintf("%s[%d]", o.Topic, o.Partition)
			value := strconv.FormatInt(o.At, 10)
			offsetsSection.Rows[i] = model.DefinitionSectionRow{Name: name, Value: value, Detail: nil}
			offsetDocs[i] = groupOffsetDoc{Name: name, Value: value}
		}
	}

	doc := struct {
		State        string `json:"state"`
		ProtocolType string `json:"protocolType"`
		Protocol     string `json:"protocol"`
		Coordinator  struct {
			Host string `json:"host"`
			Port int32  `json:"port"`
		} `json:"coordinator"`
		Members []groupMemberDoc `json:"members"`
		Offsets []groupOffsetDoc `json:"offsets"`
	}{
		State: group.State, ProtocolType: group.ProtocolType, Protocol: group.Protocol,
		Members: memberDocs, Offsets: offsetDocs,
	}
	doc.Coordinator.Host = group.Coordinator.Host
	doc.Coordinator.Port = group.Coordinator.Port
	statement, err := json.MarshalIndent(doc, "", "  ")
	if err != nil {
		return model.ObjectDefinition{}, mapError(err)
	}

	return model.ObjectDefinition{
		Path:           model.EncodePath([]model.PathSegment{{Kind: "consumerGroup", Name: groupID}}),
		Kind:           "consumerGroup",
		QualifiedName:  groupID,
		Language:       "json",
		Statements:     []string{string(statement)},
		Origin:         "server",
		Notes:          notes,
		Constraints:    []model.ConstraintMeta{},
		DocumentSchema: nil,
		Sections:       []model.DefinitionSection{groupSection, membersSection, offsetsSection},
		GeneratedAt:    time.Now().UTC().Format(time.RFC3339Nano),
	}, nil
}
